src/loan.rs

```diff
@@ -60,21 +60,177 @@ static COVENANT_PK: &str = "03b9b6059008e3576aad58e05a3a3e37133b05f68cda8535ec09
 /// `covenant_pk`, which is only used to verify that the transaction
 /// data on the witness stack matches the transaction which triggered
 /// the call.
-fn collateral_descriptor(
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+struct CollateralContract {
+    descriptor: Descriptor<PublicKey>,
     borrower_pk: PublicKey,
     lender_pk: PublicKey,
-    timelock: u64,
-    repayment_output: TxOut,
-) -> Result<Descriptor<PublicKey>> {
-    let repayment_output = serialize(&repayment_output).to_hex();
-    let desc = Descriptor::<elements::bitcoin::PublicKey>::from_str(
-        &(format!(
-            "elcovwsh({},or_i(and_v(v:pk({}),after({})),and_v(v:pk({}),outputs_pref({}))))",
-            COVENANT_PK, lender_pk, timelock, borrower_pk, repayment_output,
-        )),
-    )?;
-
-    Ok(desc)
+    timelock: u32, // TODO: Model timelocks as u32
+}
+
+impl CollateralContract {
+    /// Fill in the collateral contract template with the provided arguments.
+    fn new(
+        borrower_pk: PublicKey,
+        lender_pk: PublicKey,
+        timelock: u64,
+        repayment_output: TxOut,
+    ) -> Result<Self> {
+        let repayment_output = serialize(&repayment_output).to_hex();
+        let descriptor = Descriptor::<elements::bitcoin::PublicKey>::from_str(
+            &(format!(
+                "elcovwsh({},or_i(and_v(v:pk({}),after({})),and_v(v:pk({}),outputs_pref({}))))",
+                COVENANT_PK, lender_pk, timelock, borrower_pk, repayment_output,
+            )),
+        )
+        .context("invalid collateral output descriptor")?;
+
+        Ok(Self {
+            descriptor,
+            borrower_pk,
+            lender_pk,
+            timelock: timelock as u32,
+        })
+    }
+
+    fn address(&self) -> Address {
+        // FIXME: AddressParams should not be hard-coded
```

Comment:
Why are we hardcoding it at this level then?

Reply:
I think we completely missed this in the last project because it never caused trouble AFAIK. I kind of don't want to include the fix with this PR since it was already there and it's orthogonal to the feature, but I can be convinced otherwise haha.

Reply:
Yep makes sense.

I think this will be a hard-blocker once we want to run things on mainnet because then you actually want to pass `Liquid` into here.
It is fine to not do in this PR but you should probably take care of this soon.

Reply:
I've opened https://github.com/comit-network/baru/issues/28.